sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -711,6 +713,57 @@ export class MessageReceiver extends LinkEntity {
     };
   }
 
+  /**
+   * Prevents us from receiving any further messages.
+   */
+  public stopReceivingMessages(): Promise<void> {
+    log.receiver(`[${this.name}] User has requested to stop receiving new messages.`);
+    this._stopReceivingMessages = true;
+
+    return this.drainReceiver();
+  }
+
+  private drainReceiver(): Promise<void> {
+    log.receiver(`[${this.name}] Receiver is starting drain.`);
+
+    const drainPromise = new Promise<void>((resolve) => {
+      if (this._receiver == null) {
+        log.receiver(`[${this.name}] Internal receiver has been removed. Not draining.`);
+        resolve();
+        return;
+      }
+
+      this._receiver.once(ReceiverEvents.receiverDrained, () => {
+        log.receiver(`[${this.name}] Receiver has been drained.`);
+        resolve();
+      });
+
+      this._receiver.drain = true;
+      // this is not actually adding another credit - it'll just
+      // cause the drain call to start.
+      this._receiver.addCredit(1);
+    });
+
+    return drainPromise;
+  }
+
+  /**
+   * Adds credits to the receiver, respecting any state that
+   * indicates the receiver is closed or should not continue
+   * to receive more messages.
+   *
+   * @param credits Number of credits to add.
+   */
+  protected addCredit(credits: number): boolean {
```

Comment:
I would recommend not creating this new method for adding credits.
This opens up two ways for someone working on the project to add credits and its easy for us to get to a state where there is a mix of both patterns. We already are at that stage in the current state of this PR already.
There are few cases in this file where the `this._receiver.addCredit` is used and few where `this.addCredit` is used.

We currently already have a check on the `this._receiver` before calling addCredit, I dont mind inlining the check on `_stopReceivingMessages` along with it

Reply:
I think we can make an incremental improvement here (as I think I've done) and then start migrating all the other bits of code over. 

Otherwise we can't make any forward progress without _large_ changes and that seems riskier.

Is it not worth merging this PR because of it? Inlining everything is why we keep ending up with duplicate code, so if we can avoid it I'd prefer to.

Reply:
Also (if it helps) @HarshaNalluru  already pointed out a few spots where I should have called my function (or I could remove my old code).

So within this file, at least, there's only a single bare .addCredit call and that one is basically required.

Reply:
Leaving in and just using it in more spots (which helps with consistency, as mentioned above)